### DISCUSSION TITLE
Reference does not adopt the type of the variable that it references with strings

### DISCUSSION THREAD

**Discussion by theotherjenkutler** [savonet/liquidsoap#2234] [User Support]
**Describe the bug**
References to strings do not adopt the type string. 

**To Reproduce**

```
#!/home/jenkutler/.opam/4.10.0/bin/liquidsoap
settings.init.force_start.set(true) #this will allow LS to start w/o an active source
settings.server.telnet.set(true) #starts telnet server

set("log.file",false)
#set("log.file.path", "/home/jenkutler/ls_logs/measures.log")
set("log.stdout",true)
set("log.level",5)

def on_telconnect(m)
    amotherfuckingstring = ""
    i = ref(amotherfuckingstring)
    i := (m)
    log.debug(i)
end
server.register(usage= "mac <mac>",
                "mac", on_telconnect)
```

**Expected behavior**
That a user can connect to the telnet server and  type the mac command, enter a string and then liquidsoap would print that string to the log. (instead I get errors about how ref is type (_) but it wants a string. We all want a string. A string is all we have ever wanted.

**Version details**
 - OS: Ubuntu 20.04.3 LTS
 - Version 2.0.2

**Install method**
Opam

**Comment by smimram** [comment]
Two things:
1. in order to get the value of a reference you need to prefix it with `!` and thus replace `i` by `!i` in your script
2. the `on_teleconnect` function needs to return a string (which is printed on the telnet once the command succeeds).

Thus, the corrected version is:
```
def on_telconnect(m)
  amotherfuckingstring = ""
  i = ref(amotherfuckingstring)
  i := (m)
  log.debug(!i);
  "Done"
end
server.register(usage= "mac <mac>", "mac", on_telconnect)
```

**Comment by theotherjenkutler** [comment]
Thank you for the help! Do i still need the "!" symbol to use the reference in other ways? for example if I wanted the reference to set the mountpoint of a stream? Something like-
`output.icecast(%mp3(samplerate = 44100, internal_quality=1),
  host = "localhost", port = 8000,
  password = "measures", mount = !i,
  radio,
  fallible = true)`

**Reply by toots** [reply]
That would work yes.

**Reply by smimram** [reply]
Yes, everywhere you want to retrieve the _value_ contained in the reference (as opposed to the reference itself) you need `!`.

**Reply by theotherjenkutler** [reply]
is there a way to make a pull request for the book and/or documentation? i would be glad to contribute since others seem to be having issues with the basics as well. the book doesn't document this in a way that is understandable to me. at the very least i could be useful with my incessant question asking.

**Reply by smimram** [reply]
Of course!
- the doc is here: https://github.com/savonet/liquidsoap/tree/main/doc/content
- the book is here: https://github.com/savonet/book

**Reply by theotherjenkutler** [reply]
Thanks. Just submitted a pull request. 

I have another question regarding nested functions. I am trying to create a script where liquidsoap waits for a telnet command before creating a stream. Are there any functions built in that can create the source and output, but not have them be active until they are manually started? Creating a source and output from within a function seems to be creating some strange problems where I am unable to use event handlers such as source.on_end() when they are declared inside of another function. I am hesitant to declare the source and output globally because I don't want them to start until after a telnet command has been received and I am not sure how to control that behavior. Can you shed any light on this? Thank you!

**Reply by smimram** [reply]
Dynamic source creation can be done, but it is not easy: see https://www.liquidsoap.info/doc-2.0.2/dynamic_sources.html

**Reply by theotherjenkutler** [reply]
This is super helpful. Thank you.

I am wondering if there is a way to add a dynamic mountpoint to the equation in the example. My question is basically this: If I add a mountpoint with a reference argument to the icecast output when it is declared, will it still be changable by the telnet server? In other words, will the instance of the output be created when the output is registered or will the parameters of the icecast output be somehow implemented the second they are declared in a way that cannot be changed? I hope this question makes sense.

**Reply by theotherjenkutler** [reply]
I think what I am actually asking for are syntax examples for the "getters" that can be used with output.icecast() for example, there is documentation for on_connection but nothing for mountpoint. So I know that it is possible, I just have no idea what syntax it will expect. I imagine something like output.icecast(mountpoint=!x) . I wish that the documentation really elaborated on examples of how the syntax works. I am willing to help with examples in the documentation but I will continue to ask embarrassing questions.

**Reply by theotherjenkutler** [reply]
Also, the example throws an error---

At ./dynamic_playlist_example.liq, line 33, char 28-40:
      list.append( [(uri,s),(uri,output)],

Error 5: this value has type
  _ * source(audio=pcm(stereo), video=none, midi=none)
  .{
    time : () -> float,
    shutdown : () -> unit,
    fallible : bool,
    skip : () -> unit,
    seek : (float) -> float,
    is_active : () -> bool,
    is_up : () -> bool,
    log : 
    {level : (() -> int?).{set : ((int) -> unit)}
    },
    self_sync : () -> bool,
    duration : () -> float,
    elapsed : () -> float,
    remaining : () -> float,
    on_track : ((([string * string]) -> unit)) -> unit,
    on_leave : ((() -> unit)) -> unit,
    on_get_ready : ((() -> unit)) -> unit,
    on_shutdown : ((() -> unit)) -> unit,
    on_metadata : ((([string * string]) -> unit)) -> unit,
    last_metadata : () -> [string * string]?,
    is_ready : () -> bool,
    id : () -> string,
    current : () -> request?,
    set_queue : ([request]) -> unit,
    add : (request) -> bool,
    queue : () -> [request],
    fetch : () -> bool,
    reload : (?uri : string?) -> unit,
    length : (() -> int)
  }
but it should be a subtype of the type of the value at ./dynamic_playlist_example.liq, line 33, char 28-40
  _ * _.{is_started : _}

**Reply by toots** [reply]
We really need to type check the doc examples.. Meanwhile, I would add a `source` type annotation where ever you have a list of sources. This is a case of the issue discussed here: https://www.liquidsoap.info/doc-2.0.2/migrating.html

> Type errors with lists of sources
> Now that sources have their own methods, the actual list of methods attached to each source can vary from one to the next. For instance, playlist has a reload method but input.http does not. This currently confuses the type checker and leads to errors that look like this:
> ```
> At script.liq, line xxx, char yyy-zzz:
> Error 5: this value has type
>   _ * source(audio=?A, video=?B, midi=?C)
>   .{
>     time : () -> float,
>     shutdown : () -> unit,
>     fallible : bool,
>     skip : () -> unit,
>     seek : (float) -> float,
>     is_active : () -> bool,
>     is_up : () -> bool,
>     log :
>     {level : (() -> int?).{set : ((int) -> unit)}
>     },
>     self_sync : () -> bool,
>     duration : () -> float,
>     elapsed : () -> float,
>     remaining : () -> float,
>     on_track : ((([string * string]) -> unit)) -> unit,
>     on_leave : ((() -> unit)) -> unit,
>     on_shutdown : ((() -> unit)) -> unit,
>     on_metadata : ((([string * string]) -> unit)) -> unit,
>     is_ready : () -> bool,
>     id : () -> string,
>     selected : (() -> source(audio=?D, video=?E, midi=?F)?)
>   }
> but it should be a subtype of the type of the value at radio.liq, line 122, char 2-21
>   _ * _.{reload : _}
> ```
> In such cases, we recommend to give a little nudge to the typechecker by using the (s:source) type annotation where a list of source is causing the issue. For instance:
> ```
> s = fallback([
>   (s1:source),
>   (s2:source),
>   (s3:source)
> ])
> ```
> This tells the type checker not to worry about the source methods and just focus on what matters, that they are actually sources.. 🙂

**Reply by theotherjenkutler** [reply]
This is interesting. Thank you. Unfortunately, the error is being thrown for the output, not the source. It is saying that the output should be a source. Either way, the :source annotation returns a syntax error. None of the examples discussed here seem to be working in any capacity. At this point, I would just like to understand why source.shutdown() doesn't work. I don't need to have this whole system for dynamic streams because the source uri is always the same. Since the source is always the same, I already know the uri to shut it down. (By the way, this would be the only bit of code that made sense to me from the onset if it worked). My problem now is that source.shutdown() isn't working at all. I am able to run the script without errors, I have a source.on_end() handler which calls source.shutdown() and though the handler is activated and seems to be executed, it doesn't shut the stream down. Any thoughts on that? Thank you again.

**Reply by theotherjenkutler** [reply]
this is a snippet of my current workaround: 

`#!/home/jenkutler/.opam/4.10.0/bin/liquidsoap
settings.init.force_start.set(true) #this will allow LS to start w/o an active source
settings.server.telnet.set(true) #starts telnet server

set("log.file",false)
#set("log.file.path", "/home/jenkutler/ls_logs/measures.log")
set("log.stdout",true)
set("log.level",5)
#playlist is set to specified folder to play randomly
myPlaylist = playlist(mode='randomize',reload=1,reload_mode="rounds", "/home/jenkutler/audio/")
#radio is the source set to the playlist
radio = myPlaylist

#esp32 must send a message over telnet formatted as:
#"mac 000000000000.mp3" the telnet server will register
#the mountpoint as 000000000000.mp3 automatically
def on_telconnect(m)
    identifierstring = ""
    i = ref(identifierstring)
    i := (m)
    log.debug(!i)

    
    output.icecast(%mp3(samplerate = 44100, internal_quality=1),
    host = "localhost", port = 8000,
    password = "measures", mount = !i,
    radio,
    fallible = true)
    "Done"
end
server.register(usage= "mac <mac>",
                "mac", on_telconnect)

#todo: metadata, writing to file, end of track to shutdown stream, file transfer with server.read // file.write

def endHandler(s, m)
    log.debug("end handler") #for debugging
    source.shutdown(radio)
    #TODO: server.read & file.write methods
end
source.on_end(radio, endHandler)`
